### PR TITLE
Cli testability improvements

### LINKED ISSUE #104: Improve testability of CLI
We can improve the testability of the CLI by adding the ability for it to take cli arguments instead of the interactive project setup we currently have.

Basically, when the program is run check if any CLI arguments were passed, if there were then use those to generate the project, otherwise use the interactive process currently implemented

### PR DESCRIPTION
**Pull request by AnthonyMichaelTDM**
fixes #104 
and enables what is proposed in #103 

- [x] replace StructOpt with clap
- [x] flesh out cli for creating projects
- [x] flesh out cli for configuring existing projects
- [x] update readme to reflect CLI changes
- [x] integrate Utoipa plugin when that gets merged into main

changes how the CLI works, so I'm going to request a review when I'm done

demo of new help messages
image

### CODE REVIEW

**Comment by AnthonyMichaelTDM** [comment]
That mess of commits is because I accidentally rebased instead of merging

File: create-rust-app_cli/src/qsync/mod.rs
Comment (AnthonyMichaelTDM) [inline create-rust-app_cli/src/qsync/mod.rs:53]: this is removed, but the functionality is preserved with hidden flags in the `configure` sub command

File: create-rust-app_cli/src/main.rs
Comment (AnthonyMichaelTDM) [inline create-rust-app_cli/src/main.rs:269]: these changes are basically asking the following:
- Was a framework set with cli arguments?
    - if so use it, otherwise are we in interactive mode?
        - if so, get a framework with the interactive TUI, otherwise panic

**Review by AnthonyMichaelTDM — COMMENTED** [review]
Maybe interactive mode should default to true?

**Comment by AnthonyMichaelTDM** [comment]
this causes 'breaking changes' with how the user interacts with the CLI (commands that worked before like `create-rust-app <my project>` don't work anymore (the previous example replaced with `create-rust-app create <my project>`).

So I'm going to wait for comment or review before merging

**Comment by Wulf** [comment]
Anthony, first of all, this is all great. 

> Maybe interactive mode should default to true?

Definitely! Since we're designing the CLI for human usage primarily, we should make the default interactive. Do you think we can preserve the `create-rust-app <project_name>` UX?

**Comment by AnthonyMichaelTDM** [comment]
Yeah, of course, I can do that later this weekOn Jan 30, 2023, at 4:52 AM, Haris ***@***.***> wrote:﻿
Anthony, first of all, this is all great.

Maybe interactive mode should default to true?

Definitely! Since we're designing the CLI for human usage primarily, we should make the default interactive. Do you think we can preserve the create-rust-app <project_name> UX?

**Comment by AnthonyMichaelTDM** [comment]
Do you want that instead of, or in addition to, the subcommand UX?

Also, if it's in addition to, should using subcommands imply they want non-interactive mode and not using subcommands imply interactive (this would make implementation much simpler)

**Comment by Wulf** [comment]
If it's easier to implement, let's leave the subcommands as-is. 

So we want to preserve:
* `create-rust-app <project-name>` should bring up the interactive project creation flow
* `create-rust-app` should bring up the interactive generation options when in a rust project

Would it be possible to see if no flags/options were provided? I can see scenarios where the user enters `create-rust-app create` but was trying to create a project named "create". We should warn on this. If this is too much work, let's skip this.

**Comment by AnthonyMichaelTDM** [comment]
Interactive mode is now the default, and the older UX is reimplemented, found out about the args_conflicts_with_subcommands option which helped streamline things a lot I think.

The nested match logic isn't ideal, but it gets the job done

here's a screenshot of what the help message looks like now 
image

File: create-rust-app_cli/src/main.rs
Comment (AnthonyMichaelTDM) [inline create-rust-app_cli/src/main.rs:219]: This is the nested match I was referring to, on second look it's not really that bad, I guess.

**Comment by AnthonyMichaelTDM** [comment]
[this commit](https://github.com/Wulf/create-rust-app/pull/109/commits/5edf5cdca5dc2b8f4e08be411b87e2813b68b1c8) also fixes what's causing *some* of the dependabot PR's to fail their CI

**Comment by AnthonyMichaelTDM** [comment]
I don't want to be stepping on any toes here so I'm going to wait for another review before merging

File: create-rust-app_cli/src/main.rs
Comment (Wulf) [inline create-rust-app_cli/src/main.rs:504]: :100:

**Review by Wulf — APPROVED** [review]
awesome, looks great @AnthonyMichaelTDM!

**Pull request merged**
